File: examples/typescript/sdk-reference/error-handling/serdes-error.ts
import { withDurableExecution, Serdes, SerdesContext } from "@aws/durable-execution-sdk-js";

// A custom Serdes that throws on serialization failure.
// The SDK terminates the Lambda invocation when serialization fails.
const strictSerdes: Serdes<unknown> = {
  serialize: async (value: unknown, _context: SerdesContext) => {
    // Circular references cause JSON.stringify to throw
    return JSON.stringify(value);
  },
  deserialize: async (data: string | undefined, _context: SerdesContext) => {
    if (data === undefined) return undefined;
    return JSON.parse(data);
  },
};

export const handler = withDurableExecution(async (event, context) => {
  const result = await context.step(
    "build-result",
    async (stepCtx) => {
      // Return a value that can be serialized
      return { message: "hello", timestamp: Date.now() };
    },
    { serdes: strictSerdes },
  );
  return result;
});

File: examples/csharp/sdk-reference/error-handling/serdes-error.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;
using Microsoft.Extensions.Logging;

public class SerdesErrorExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<object, ResultData>(Workflow, input, context);

    private async Task<ResultData> Workflow(object input, IDurableContext ctx)
    {
        // The step result is serialized to checkpoint storage with the
        // ILambdaSerializer registered on ILambdaContext.Serializer. When
        // serialization or deserialization fails, the SDK surfaces the failure as
        // an unhandled exception, returns a FAILED status, and does not retry.
        try
        {
            ResultData result = await ctx.StepAsync(
                async (_, _) => new ResultData("hello"),
                name: "build-result");
            return result;
        }
        catch (Exception e)
        {
            ctx.Logger.LogError("Serialization failed: {Message}", e.Message);
            throw;
        }
    }
}

public record ResultData(string Message);
